micro-rdk/src/common/conn/viam.rs

```diff
@@ -0,0 +1,1723 @@
+
+            // cancel the fake app task
+            // this should simulate a network loss or app.viam.com going offline for some
+            // reasons. We should still be able to make a connection to the H2 server
+            assert!(fake_server_task.cancel().await.is_none());
+            let _ = Timer::after(Duration::from_millis(300));
+            let t2 = test_connect_to(addr, cloned_exec.clone()).await;
+            assert!(t2.is_ok());
+            assert_eq!(shared_auth_counter.load(Ordering::Acquire), 1);
+
+            let other_clone = cloned_exec.clone();
+            // bring up a new app client to simulate network resuming.
+            // we just check that another call to authenticate has been issues
+            // this indicate that a connection was made.
+            let _fake_server_task = cloned_exec
+                .spawn(async move { run_fake_app_server(other_clone, cloned_app).await });
+            let _ = Timer::after(Duration::from_secs(2)).await;
+            assert_eq!(shared_auth_counter.load(Ordering::Acquire), 2);
+        });
+    }
+
+    #[test_log::test]
+    /// Runs viam server exposing HTTP2 connections, since each HTTP2 connection gets a
+    /// max_prio assigned we can't test preemption
+    /// Testing webrtc would require to add support for ice control agent931
+    fn test_multiple_connection_http2() {
+        let _unused = global_network_test_lock();
+        let ram_storage = RAMStorage::new();
+        let network = match local_ip_address::local_ip().expect("error parsing local IP") {
+            std::net::IpAddr::V4(ip) => ExternallyManagedNetwork::new(ip),
+            _ => panic!("oops expected ipv4"),
+        };
+
+        let creds = CloudConfig {
+            id: "".to_string(),
+            secret: "".to_string(),
+            app_address: "".to_owned(),
+        };
+
+        assert!(ram_storage.store_robot_credentials(creds).is_ok());
+
+        let mdns = NativeMdns::new("".to_owned(), network.get_ip());
+        assert!(mdns.is_ok());
+        let mdns = mdns.unwrap();
+
+        let mut viam_server = ViamServerBuilder::new(ram_storage);
+        viam_server
+            .with_app_uri("http://localhost:56563".try_into().unwrap())
+            .with_http2_server(NativeH2Connector::default(), 12346)
+            .with_max_concurrent_connection(3);
+
+        let exec = Executor::new();
+
+        let mut viam_server = viam_server.build(
+            NativeH2Connector::default(),
+            exec.clone(),
+            mdns,
+            Box::new(network),
+        );
+        let cloned_exec = exec.clone();
+
+        let mut app = AppServerInsecure::default();
+        app.config_fn = Some(Rc::new(Box::new(|| {
+            let mut cfg = make_sample_config();
+            if let Some(cloud) = cfg.cloud.as_mut() {
+                cloud.fqdn = "test-bot.xxds65ui.viam.cloud".to_owned();
+                cloud.local_fqdn = "test-bot.xxds65ui.viam.local.cloud".to_owned();
```

Comment:
can we add a TODO to have these test addresses centralized or come a dev config?